src/idpyoidc/server/session/database.py

```diff
@@ -44,153 +38,147 @@ def __init__(self, crypt_config: Optional[dict] = None, **kwargs):
+
+    @staticmethod
+    def unpack_branch_key(key):
+        """ Translate a key into an ordered list of names """
+        return key.split(DIVIDER)
+
+    def encrypted_branch_id(self, *args) -> str:
+        """ Provided an ordered list of names construct a key and then encrypt it. """
+        rnd = rndstr(32)
+        return base64.b64encode(
+            self.crypt.encrypt(lv_pack(rnd, self.branch_key(*args)).encode())
+        ).decode("utf-8")
+
+    def decrypt_branch_id(self, key: str) -> List[str]:
         """
+            Given an encrypted key, decrypt it and then unpack the key to return an ordered list
+            of names.
+        """
+        try:
+            plain = self.crypt.decrypt(base64.b64decode(key))
+        except cryptography.fernet.InvalidToken as err:
+            logger.error(f"cryptography.fernet.InvalidToken: {key}")
+            raise ValueError(err)
+        except Exception as err:
+            raise ValueError(err)
+        # order: rnd, type, sid
+        return self.unpack_branch_key(lv_unpack(as_unicode(plain))[1])
 
-        :param path: a list of identifiers
+    def set(self, path: List[str], value: Union[NodeInfo, Grant]):
+        """
+        Assign a value to an node in the database.
+        As a side effect create a list of nodes (the branch) leading up to the leaf node.
+
+        :param path: a list of identifiers. root -> .. -> leaf
         :param value: Class instance to be stored
         """
 
-        uid, client_id, grant_id = self._eval_path(path)
-
-        if grant_id:
-            gid_key = self.session_key(uid, client_id, grant_id)
-            self.db[gid_key] = value
-
-        if client_id:
-            cid_key = self.session_key(uid, client_id)
-            cid_info = self.db.get(cid_key, ClientSessionInfo())
-            if not grant_id:
-                self.db[cid_key] = value
-            elif grant_id not in cid_info.subordinate:
-                cid_info.add_subordinate(grant_id)
-                self.db[cid_key] = cid_info
-
-        userinfo = self.db.get(uid, UserSessionInfo())
-        if client_id is None:
-            self.db[uid] = value
-        if client_id and client_id not in userinfo.subordinate:
-            userinfo.add_subordinate(client_id)
-            self.db[uid] = userinfo
-
-    def get(self, path: List[str]) -> Union[SessionInfo, Grant]:
-        uid, client_id, grant_id = self._eval_path(path)
-        try:
-            user_info = self.db[uid]
-        except KeyError:
-            raise KeyError("No such UserID")
-        except TypeError:
-            raise InconsistentDatabase("Missing session db")
-        else:
-            if user_info is None:
-                raise KeyError("No such UserID")
+        _len = len(path)
+
+        _superior = None
+        for i in range(_len):
+            _key = self.branch_key(*path[0:i + 1])
+            # _key = path[i]
+            _info = self.db.get(_key)
+            if _info is None:
```

Comment:
if not _info?

Reply:
Well ! Each node are expected to be an instance of a class. No other values should be there.
A instance of class with no attributes may trigger `not _info` but not `_info is None`.